osdu/services/authentication.py

```diff
@@ -0,0 +1,68 @@
+""" Provides a simple Python interface to the OSDU Authentication API.
+"""
+import requests
+from .base import BaseService
+from time import time
+
+class AuthenticationService(BaseService):
+
+    def update_token(client):
+        """Determines if the current access token associated with the client has expired.
+        If the token is not expired, the current access_token will be returned, unchanged.
+        If the token has expired, this function will attempt to refresh it, update it on client, and return it.
+        For simple clients, refresh requires a OSDU_CLIENTWITHSECRET_ID, OSDU_CLIENTWITHSECRET_SECRET, REFRESH_TOKEN, and REFRESH_URL
+        For Service Principal clients, refresh requires a resource_prefix and AWS_PROFILE (same as initial auth)
+        For AWS clients, refresh requires OSDU_USER, OSDU_PASSWORD, AWS_PROFILE, and OSDU_CLIENT_ID
+        
+        :param client: client in use
+
+        :returns: tuple containing 2 items: the new access token and it's expiration time
+                        - access_token: used to access OSDU services
+                        - expires_in:   expiration time for the token
+        """
+        if(AuthenticationService._need_update_token(client)):
+            if (hasattr(client, "resource_prefix") and client.resource_prefix is not None): #service principal client
+                token = AuthenticationService._update_token_service_principal(client)
+            elif (hasattr(client, "profile") and client.profile is not None): #aws client
+                token = AuthenticationService._update_token_aws(client)
+            else: #simple client
+                token = AuthenticationService._update_token_simple(client)
+        else:
+            token = client.access_token, client._token_expiration if hasattr(client, "_token_expiration") else None
+        return token
+    
+    def _need_update_token(client):
+        return hasattr(client, "_token_expiration") and client._token_expiration < time() or client.access_token is None
+
+    
+    def _update_token_simple(client) -> dict:
```

Comment:
We should only need a single `update_token` method that would be implemented for each client according to that client's auth needs. Since there is currently a 1:1 relationship between the 3 clients and the 3 methods of retrieving and updating the token, I don't know that there is a clear need yet for a separate `AuthorizatinoService`, though I'm open to any persuasive arguments for it :)

Reply:
AuthenticationService is no more!